### PR TITLE
Centerpoint config detector

### PR DESCRIPTION
**Pull request by yinchimaoliang**
Add configs and centerpoint detector.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Please rename the configs following the [config name style](https://mmdetection3d.readthedocs.io/en/latest/config.html#config-name-style).

File: configs/_base_/models/centerpoint_01voxel_second_secfpn_nus.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        in_channels=sum([256, 256]),
+        tasks=[
+            dict(num_class=1, class_names=['car']),
+            dict(num_class=2, class_names=['truck', 'construction_vehicle']),
+            dict(num_class=2, class_names=['bus', 'trailer']),
+            dict(num_class=1, class_names=['barrier']),
+            dict(num_class=2, class_names=['motorcycle', 'bicycle']),
+            dict(num_class=2, class_names=['pedestrian', 'traffic_cone']),
+        ],
+        common_heads={
```
Comment (ZwwWayne) [inline configs/_base_/models/centerpoint_01voxel_second_secfpn_nus.py:49]:
```python
common_heads=dict(
            reg=(2, 2),
            height=(1, 2),
            dim=(3, 2),
            rot=(2, 2),
            vel=(2, 2))

```

File: configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
+_base_ = [
+    '../_base_/models/centerpoint_01voxel_second_secfpn_nus.py',
+    '../_base_/default_runtime.py'
+]
+
+# If point cloud range is changed, the models should also change their point
+# cloud range accordingly
+voxel_size = [0.075, 0.075, 0.2]
+point_cloud_range = [-54, -54, -5.0, 54, 54, 3.0]
```
Comment (ZwwWayne) [inline configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py:9]: `voxel_size` and `point_cloud_range` are middle variables. We also need to change those keys in model to really modify them.

File: configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
+_base_ = [
+    '../_base_/models/centerpoint_01voxel_second_secfpn_nus.py',
+    '../_base_/default_runtime.py'
```
Comment (ZwwWayne) [inline configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py:3]: We can inherit `nus-3d.py` as `base` and skip some duplicated keys such as path and data_root.

File: mmdet3d/models/detectors/centerpoint.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                 test_cfg=None,
+                 pretrained=None):
+        super(CenterPoint,
+              self).__init__(pts_voxel_layer, pts_voxel_encoder,
+                             pts_middle_encoder, pts_fusion_layer,
+                             img_backbone, pts_backbone, img_neck, pts_neck,
+                             pts_bbox_head, img_roi_head, img_rpn_head,
+                             train_cfg, test_cfg, pretrained)
+
+    def extract_pts_feat(self, pts, img_feats, img_metas):
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/centerpoint.py:32]: Is this function necessary?

File: mmdet3d/models/detectors/centerpoint.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    def simple_test_pts(self, x, img_metas, rescale=False):
+        """Test function of point cloud branch."""
+        outs = self.pts_bbox_head(x)
+        bbox_list = self.pts_bbox_head.get_bboxes(
+            outs, img_metas, rescale=rescale)
+        bbox_results = [
+            bbox3d2result(bboxes, scores, labels)
+            for bboxes, scores, labels in bbox_list
+        ]
+        return bbox_results[0]
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/centerpoint.py:81]: There should not contain [0]

File: configs/centerpoint/README.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  journal={arXiv preprint arXiv:2006.11275},
+  year={2020}
+}
+```
+
+## Results
+
+### CenterPoint
+
+|Backbone|  Voxel type   |Dcn|Circular nms| Mem (GB) | Inf time (fps) | mAP |NDS| Download |
```
Comment (ZwwWayne) [inline configs/centerpoint/README.md:40]: Voxel type -> Voxel type (voxel size)

File: configs/centerpoint/README.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+| :---------: |:-----: |:-----: | :------: | :------------: | :----: |:----: | :------: |:------: |
+|[SECFPN](./centerpoint_01voxel_second_secfpn_4x8_cyclic_20e_nus.py)|01voxel|✗|✗||||||
```
Comment (ZwwWayne) [inline configs/centerpoint/README.md:42]: 01voxel -> voxel (0.1), also change this for other configs.

File: mmdet3d/models/detectors/centerpoint.py (hunk 4 above)
Comment (yinchimaoliang) [reply mmdet3d/models/detectors/centerpoint.py:32]: Yes, we can't use this function from mvx_two_stage.

File: configs/_base_/models/centerpoint_02pillar_second_secfpn_nus.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    type='CenterPoint',
+    pts_voxel_layer=dict(
+        max_num_points=20, voxel_size=voxel_size, max_voxels=(30000, 40000)),
+    pts_voxel_encoder=dict(
+        type='PillarFeatureNet',
+        in_channels=5,
+        feat_channels=[64],
+        with_distance=False,
+        voxel_size=(0.2, 0.2, 8),
+        norm_cfg=dict(type='BN1d', eps=1e-3, momentum=0.01),
```
Comment (ZwwWayne) [inline configs/_base_/models/centerpoint_02pillar_second_secfpn_nus.py:12]: Should we use BN or SyncBN here?

File: configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+# For nuScenes we usually do 10-class detection
+class_names = [
+    'car', 'truck', 'construction_vehicle', 'bus', 'trailer', 'barrier',
+    'motorcycle', 'bicycle', 'pedestrian', 'traffic_cone'
+]
+dataset_type = 'NuScenesDataset'
+data_root = 'data/nuscenes/'
+# Input modality for nuScenes dataset, this is consistent with the submission
+# format which requires the information in input_modality.
+input_modality = dict(
```
Comment (ZwwWayne) [inline configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py:16]: The key input_modality is unnecessary here, remove it for all configs in configs/centerpoint/

File: configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+            dict(type='Collect3D', keys=['points'])
+        ])
+]
+
+data = dict(
+    train=dict(dataset=dict(pipeline=train_pipeline)),
+    val=dict(pipeline=test_pipeline),
+    test=dict(pipeline=test_pipeline))
+
+model = dict(
```
Comment (ZwwWayne) [inline configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py:125]: Move the config of model, train_cfg and test_cfg to the top below base and above data.

File: configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+            car=5,
+            truck=5,
+            bus=5,
+            trailer=5,
+            construction_vehicle=5,
+            traffic_cone=5,
+            barrier=5,
+            motorcycle=5,
+            bicycle=5,
+            pedestrian=5,
```
Comment (ZwwWayne) [inline configs/centerpoint/centerpoint_0075voxel_second_secfpn_4x8_cyclic_20e_nus.py:50]: delete the unnecessary comma

File: configs/centerpoint/centerpoint_01voxel_second_secfpn_4x8_cyclic_20e_nus.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    val=dict(pipeline=test_pipeline, classes=class_names),
+    test=dict(pipeline=test_pipeline, classes=class_names))
+
+# For nuScenes dataset, we usually evaluate the model at the end of training.
+# Since the models are trained by 24 epochs by default, we set evaluation
+# interval to be 24. Please change the interval accordingly if you do not
+# use a default schedule.
+# optimizer
+# This schedule is mainly used by models on nuScenes dataset
+optimizer = dict(type='AdamW', lr=1e-4, weight_decay=0.01)
```
Comment (ZwwWayne) [inline configs/centerpoint/centerpoint_01voxel_second_secfpn_4x8_cyclic_20e_nus.py:152]: We can still use `cyclic_40e` schedule and only change `total_epochs`, `optimizer_config`, and `optimizer`, or create a cyclic_20e.py in `__base__` and use that for all centerpoint models.

File: configs/_base_/schedules/cyclic_20e.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# For nuScenes dataset, we usually evaluate the model at the end of training.
+# Since the models are trained by 24 epochs by default, we set evaluation
```
Comment (ZwwWayne) [inline configs/_base_/schedules/cyclic_20e.py:2]: Modify comments

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**